standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/RetryingHMSHandler.java

```diff
@@ -61,6 +66,7 @@ public Result(Object result, int numRetries) {
 
   private final Configuration origConf;            // base configuration
   private final Configuration activeConf;  // active configuration
+  private final List<MetaStoreEndFunctionListener> endFunctionListeners; // the end function listener
```

Comment:
didn't know we had something like this :)

but having only `MetaStoreEndFunctionListener` seems a bit odd;

how about:
* introduce a `MetaStoreFunctionListener`
* keep the `MetaStoreEndFunctionListener` interface but extend the new interface with default implementation of `startFunction`
* move the old logging/counting stuff into a `MetaStoreFunctionListener`

just an idea; let me know what you think!

Reply:
Thank you for the comments!
The `startFunctions` could be eliminated for the reasons listing above, leaving the `MetaStoreEndFunctionListener` on `endFunction` that should be taken care of, but introducing a `MetaStoreFunctionListener` makes things much better and straightforward. A user can monitor the input or output of the method, or implement his own audit logics.
This change also introduce some incompatibility in [MetaStoreEndFunctionContext](https://github.com/apache/hive/blob/master/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreEndFunctionContext.java#L55-L57), the `getInputTableName` would return null for the old implemention.